### PR TITLE
threads: timeout hash table

### PR DESCRIPTION
**Pull request by slavidichkov**

### CODE REVIEW

File: src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    public TimeoutRemover(TimeoutHashtable table, String key, long timeOut) {
+        this.table = table;
+        this.key = key;
+        this.timeOut = timeOut;
+    }
+    @Override
+    public void run() {
+        try {
+            this.sleep(timeOut);
```
Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java]: By using this way, you are not able to reset counter. You always have to create a new thread which is an expensive operation.

Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java]: done

File: src/com/clouway/task5/timeouthastablelast/Demo.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package com.clouway.task5.timeouthastablelast;
+
+/**
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public class Demo {
+    public static void main(String[] args) throws InterruptedException {
+        TimeoutHashtable hashtable=new TimeoutHashtable(1000);
+
+        hashtable.put("slavi","slavi");
```
Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/Demo.java]: As I can see you made it to be only for string keys and values. Make it generic !

Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/Demo.java]: I made my TimeoutHashtable generic

Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/Demo.java]: As I can see, the key is still always string... It should be generic type too.

Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/Demo.java]: done

File: src/com/clouway/task5/timeouthastablelast/Demo.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package com.clouway.task5.timeouthastablelast;
+
+import java.util.Date;
+
+/**
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public class Demo {
+    public static void main(String[] args) throws InterruptedException {
+        TimeoutHashtable<String,Date> hashtable=new TimeoutHashtable(1000);
```
Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/Demo.java]:
```java
new TemeoutHashtable<String, Date>(1000);
// for java >= 1.7
new TemeoutHashtable<>(1000);

```

File: src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    private final long timeOut;
+
+    public TimeoutHashtable(long timeOut) {
+        this.timeOut = timeOut;
+    }
+
+    public void put(K key, V value) {
+        if (elementsTabel.containsKey(key)) {
+            hashtableCleaner.restart();
+            elementsTabel.remove(key);
```
Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java]: There is a typo here: elementsTabel

Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java]: Why you are are removing element from the hashtable? You have to override the existing value I suppose.

File: src/com/clouway/task5/timeouthastablelast/Demo.java (hunk 3 above)
Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/Demo.java]: I missed it ...

Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/Demo.java]: done

File: src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package com.clouway.task5.timeouthastablelast;
+
+/**
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public class TimeoutRemover<K> extends Thread {
+    private TimeoutHashtable table;
```
Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java]: TimeoutHashtable is defined as:
```java
class TimeoutHashtable <K,V>
```

File: src/com/clouway/task5/timeouthastablelast/HashtableCleaner.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
+package com.clouway.task5.timeouthastablelast;
+
+/**
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public class HashtableCleaner<T> {
```
Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/HashtableCleaner.java]: I'm not sure that you need this class, probably you have to consider to remove it.

Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/HashtableCleaner.java]: done

File: src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java (hunk 5 above)
Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java]: done

File: src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java (hunk 4 above)
Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java]: done

File: src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    @Override
+    public void run() {
+        try {
+            while (counter<timeOut){
+                sleep(1);
+                counter++;
+            }
+            table.remove(key);
+        } catch (InterruptedException e) {
+
```
Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java]: On interrupt you have to remove key, cause this item will never be removed.

File: src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public class TimeoutRemover<K,V> extends Thread {
+    private TimeoutHashtable<K,V> table;
+    private final K key;
+    private V value;
+    private long timeOut;
+    private long counter;
+
+    public void setCounter(long counter) {
```
Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java]: This method is not used.

File: src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.Hashtable;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public class TimeoutHashtable<K,V> {
+    private Hashtable<K, TimeoutRemover> elementsTabel = new Hashtable<K, TimeoutRemover>();
+    private TimeoutRemover<K,V> timeoutRemover;
```
Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java]: I'm not sure that you need this class level reference.

File: src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java (hunk 8 above)
Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java]: I forgot to delete it

Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java]: Make sure that you looked at your code, before "waiting for review" phase. You are wasting time of the reviewer that way.

File: src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java (hunk 7 above)
Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java]: done

File: src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java (hunk 9 above)
Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java]: I use it in my all methods ....

File: src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java (hunk 8 above)
Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutRemover.java]: OK

File: src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java (hunk 9 above)
Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java]: This doesn't make it right. 

When you add new entry to the hashtable, you are always creating a new instance of TimeoutRemover. If so, why you are not having list of TimeoutRevers, but you keeping only one? Why you have to keep reference to the remover when you already have it from the map value ?

File: src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+import java.util.Hashtable;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public class TimeoutHashtable<K,V> {
+    private Hashtable<K, TimeoutRemover> elementsTabel = new Hashtable<K, TimeoutRemover>();
```
Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java]: TimeoutRemover is defined as 
```
<K, V>
```

File: src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
+package com.clouway.task5.timeouthastablelast;
+
+import java.util.Hashtable;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public class TimeoutHashtable<K,V> {
```
Comment (mgenov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java]: TimeoutRemover is defined as 
```
<K, V>
```

Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java]: yes

File: src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java (hunk 9 above)
Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java]: fixed

File: src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java (hunk 11 above)
Comment (slavidichkov) [inline src/com/clouway/task5/timeouthastablelast/TimeoutHashtable.java]: fixed

**Comment by mgenov** [comment]
LGTM !

**Pull request merged**